File: Problem-13.py
"""
Time complexity: O(n)
Space complexity: O(n)
Approach: Maintain a hashmap, mapping each unique letter in s to each unique letter in t.
            Maintain a set to mark all the unique letters in t that have been already mapped.
            When a new letter in s is mapped to a letter different from one in the hashmap or
            if a new letter is mapped to a letter in t that is already mapped to a different letter
            then we return false
            else we return true.

"""
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        smap = {}
        tset = set()
        for i in range(len(s)):
            if s[i] in smap:
                if smap[s[i]] != t[i]:
                    return False
            else:
                if t[i] in tset:
                    return False
                else:
                    smap[s[i]] = t[i]
                    tset.add(t[i])
        return True

File: Problem_13.java
import java.util.HashMap;

class Solution {
    /*
     * Time Complexity - O(n) where n is the size of string 
     * Space Complexity - O(1)
     */
    public boolean isIsomorphic(String s, String t) {
        HashMap<Character, Character> sMap = new HashMap<Character, Character>();
        HashMap<Character, Character> tMap = new HashMap<Character, Character>();
        for (int i=0; i<s.length(); i++) {
            char sChar = s.charAt(i);
            char tChar = t.charAt(i);
            if (sMap.containsKey(sChar)) {
                if (sMap.get(sChar) != tChar)
                    return false;
            } else {
                sMap.put(sChar, tChar);
            }
            if (tMap.containsKey(tChar)) {
                if (tMap.get(tChar) != sChar)
                    return false;
            } else {
                tMap.put(tChar, sChar);
            }
        }
        return true;
    }
}